smart-contracts/assembly/contracts/NFT/NFT-internals.ts

```diff
@@ -1,282 +1,231 @@
-
-// ==================================================== //
-// ====                 TRANSFER                   ==== //
-// ==================================================== //
-
-export function _transfer(
-  caller: string,
+function operatorAllowanceKey(
   owner: string,
-  recipient: string,
-  tokenId: u64,
-): void {
-  assertIsMinted(tokenId);
-  assertIsOwner(owner, tokenId);
-  assertNotSelfTransfer(owner, recipient);
-  assertIsApproved(owner, caller, tokenId);
-
-  _removeApproval(tokenId);
-  _updateBalanceOf(owner, false);
-  _updateBalanceOf(recipient, true);
-
-  Storage.set(ownerTokenKey + tokenId.toString(), recipient);
+  operator: string,
+): StaticArray<u8> {
+  return OPERATOR_ALLOWANCE_KEY_PREFIX.concat(
+    stringToBytes(owner).concat(stringToBytes(operator)),
+  );
 }
 
-// ==================================================== //
-// ====                 APPROVAL                   ==== //
-// ==================================================== //
-
 /**
- * Store the approved address for a token
+ * Count all NFTs assigned to an owner.
  *
- * @param owner - owner address
- * @param spenderAddress - spender address
- * @param tokenId - The token ID to approve
+ * @param owner - An address for whom to query the balance
  */
-export function _approve(
-  tokenId: u64,
-  owner: string,
-  spenderAddress: string,
-): void {
-  assertIsMinted(tokenId);
-  assertIsOwner(owner, tokenId);
-  assert(!_isApproved(spenderAddress, tokenId), 'Already approved');
-
-  const key = approvedTokenKey + tokenId.toString();
-  Storage.set(key, spenderAddress);
+export function _balanceOf(owner: string): u256 {
+  const key = balanceKey(owner);
+  if (Storage.has(key)) {
+    return bytesToU256(Storage.get(key));
+  }
+  return u256.Zero;
 }
 
 /**
- * Removes the approval of the token
- * @param tokenId - the tokenID
+ * Returns the owner of a given tokenId.
+ *
+ * @param tokenId - The identifier for an NFT
+ * @returns the address of the owner of the NFT
  */
-function _removeApproval(tokenId: u64): void {
-  const key = approvedTokenKey + tokenId.toString();
-  Storage.set(key, '');
+export function _ownerOf(tokenId: u256): string {
+  const key = ownerKey(tokenId);
+  if (Storage.has(key)) {
+    return bytesToString(Storage.get(key));
+  }
+  return '';
 }
 
-export function _getApproved(tokenId: u64): string {
-  const key = approvedTokenKey + tokenId.toString();
-  if (!Storage.has(key)) return '';
-
-  return Storage.get(key);
+/**
+ * Returns the name of the contract.
+ */
+export function _name(): string {
+  if (Storage.has(NAME_KEY)) {
+    return bytesToString(Storage.get(NAME_KEY));
+  }
+  return '';
 }
 
-export function _isApproved(address: string, tokenId: u64): bool {
-  if (address.length === 0) return false;
-  const approvedAddress = _getApproved(tokenId);
-  return approvedAddress === address;
+/**
+ * Returns the symbol of the contract.
+ */
+export function _symbol(): string {
+  if (Storage.has(SYMBOL_KEY)) {
```

Comment:
useless if considering this is set in contructor. isnt it ?

Reply:
It is indeed set in the constructor but I have no guarantee that final contract will call the super constructor, that is why I added this verification. But if you think it's irrelevant and that developers should make sure to call the super constructor then I can remove it.